nexus/src/authz/oso_generic.rs

```diff
@@ -0,0 +1,170 @@
+
+//! Oso integration
+
+use super::actor::AnyActor;
+use super::actor::AuthenticatedActor;
+use super::api_resources::Fleet;
+use super::api_resources::FleetChild;
+use super::api_resources::Organization;
+use super::api_resources::Project;
+use super::api_resources::ProjectChild;
+use super::roles::AuthzResource;
+use super::roles::RoleSet;
+use crate::authn;
+use crate::context::OpContext;
+use crate::db::DataStore;
+use anyhow::Context;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use omicron_common::api::external::Error;
+use oso::Oso;
+use oso::PolarClass;
+use std::fmt;
+
+/// Polar configuration describing control plane authorization rules
+pub const OMICRON_AUTHZ_CONFIG: &str = include_str!("omicron.polar");
+
+/// Returns an Oso handle suitable for authorizing using Omicron's authorization
+/// rules
+pub fn make_omicron_oso() -> Result<Oso, anyhow::Error> {
+    let mut oso = Oso::new();
+    // TODO-cleanup There is a lot of boilerplate in the definitions of these
+    // structures as they relate to Polar.  For example, most of them impl Eq
+    // and PartialEq, and the corresponding PolarClass should have an equality
+    // impl as well.  The resources all have a "has_role" Polar method that all
+    // do the same thing.  It'd be nice to find a way to commonize these.  A
+    // macro might help.
+    let classes = [
+        Action::get_polar_class(),
+        AnyActor::get_polar_class(),
+        AuthenticatedActor::get_polar_class(),
+        Database::get_polar_class(),
+        Fleet::get_polar_class(),
+        Organization::get_polar_class(),
+        Project::get_polar_class(),
+        ProjectChild::get_polar_class(),
+        FleetChild::get_polar_class(),
+    ];
+    for c in classes {
+        oso.register_class(c).context("registering class")?;
+    }
+    oso.load_str(OMICRON_AUTHZ_CONFIG)
+        .context("loading built-in Polar (Oso) config")?;
+    Ok(oso)
+}
+
+/// Describes an action being authorized
+///
+/// There's currently just one enum of Actions for all of Omicron.  We expect
+/// most objects to support mostly the same set of actions.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Action {
+    Query, // only used for [`Database`]
+    Read,
+    Modify,
+    Delete,
+    ListChildren,
+    CreateChild,
+}
+
+impl oso::PolarClass for Action {
+    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
+        oso::Class::builder().with_equality_check().add_method(
+            "to_perm",
+            |a: &Action| {
+                match a {
+                    Action::Query => Perm::Query,
+                    Action::Read => Perm::Read,
+                    Action::Modify => Perm::Modify,
+                    Action::Delete => Perm::Modify,
+                    Action::ListChildren => Perm::ListChildren,
+                    Action::CreateChild => Perm::CreateChild,
+                }
+                .to_string()
+            },
+        )
+    }
+}
```

Comment:
Does this basically answer the question "what permission do I need to perform this action?" If so, does that mean only one permission can authorize a particular action, or could this be done in a way that actions map to multiple permissions?

Reply:
In the Oso docs I detect a little ambiguity. Sometimes talk as though actions and permissions are the same, or at least match up one to one. E.g., second argument to `has_permission` is called an action here:

```prolog
allow(actor, action, resource) if
  has_permission(actor, action, resource);
```

https://docs.osohq.com/guides/rbac.html#declare-roles-and-permissions

Reply:
> Does this basically answer the question "what permission do I need to perform this action?"

Yes.

> In the Oso docs I detect a little ambiguity. Sometimes they talk as though actions and permissions are the same, or at least match up one to one.

Yeah, I can't say I understand the intended distinction between actions and permissions.  As you said: the thing seems to come in as an "action" and be used as a "permission", and all the examples I've seen are strings.

I wrote this (tiny) abstraction layer to try to make that a bit more explicit.  Nexus at-large authorizes you for an _action_ and we translate it (mostly 1-1) to a _permission_ and use permissions in the policy.

> If so, does that mean only one permission can authorize a particular action, or could this be done in a way that actions map to multiple permissions?

I think the way this is set up today, that's true.  In principle, we could have `to_perm()` return a list of permissions and the policy could check all of them separately.  In practice, I think that sort of thing is generally discouraged.

Reply:
I see, that makes sense. Could you get rid of Perm entirely then and impl `Display` directly on `Action` instead of converting `Action` to `Perm` and then calling `to_string` on the latter?

Related, why is there a delete action but no delete permission?

Reply:
I see now they are called `permissions` in the policy, so I take it the `Perm` thing is meant to make clear that what we're calling Actions elsewhere become permissions once we're inside the policy.

Reply:
> Could you get rid of Perm entirely then and impl Display directly on Action instead of converting Action to Perm and then calling to_string on the latter?

I think we could, but I think that conflates things (the display name of an Action does not need to be exactly the same string that we use all over the policy file).  It feels like it makes it easier to confuse an arbitrary String with a Permission.

> Related, why is there a delete action but no delete permission?

Short answer: because there's no role in the policy that would have "delete" and not "modify", so there's no point in distinguishing them in the policy.  (It _is_ useful to distinguish the actions, and that's the one case where actions don't map 1-1 with permissions.)

Basically I wrote the existing policy (i.e., the contents of omicron.polar) based on concrete use cases and I didn't see an obvious use case for someone being able to delete a thing but not be able to modify it.

Reply:
> I see now they are called `permissions` in the policy, so I take it the `Perm` thing is meant to make clear that what we're calling Actions elsewhere become permissions once we're inside the policy.

Yeah that's a much better way to put it than what I said!

Reply:
> I wouldn't mind a line explaining this in the Perm doc comment.

I've done this in 7020feba03be14f016bea5e629c956c8c9096c34.  I also moved the guts of the conversion out of the "to_perm" Polar method and into an explicit `impl From<&Action> for Perm`.  To me, that more clearly communicates that `Perm`'s job is to aid in this translation.